core/src/main/java/org/alfresco/httpclient/HttpClient4Factory.java

```diff
@@ -0,0 +1,104 @@
+ * GNU Lesser General Public License for more details.
+ * -
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Alfresco. If not, see <http://www.gnu.org/licenses/>.
+ * #L%
+ */
+
+package org.alfresco.httpclient;
+
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManager;
+
+import org.alfresco.error.AlfrescoRuntimeException;
+import org.apache.http.HttpHost;
+import org.apache.http.HttpRequestInterceptor;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.conn.ssl.SSLConnectionSocketFactory;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.impl.client.HttpClients;
+
+
+public class HttpClient4Factory
+{
+    protected static final String TLS_PROTOCOL = "TLS";
+    protected static final String HTTPCLIENT_CONFIG = "httpclient.config.";
+    protected static final String HTTPS_PROTOCOL = "https";
+    protected static final String HTTP_TARGET_HOST = "http.target_host";
+    protected static final String TLS_V_1_2 = "TLSv1.2";
+    protected static final String TLS_V_1_3 = "TLSv1.3";
+
+    private static SSLContext createSSLContext(HttpClientConfig config)
+    {
+        KeyManager[] keyManagers = config.getKeyStore().createKeyManagers();
+        TrustManager[] trustManagers = config.getTrustStore().createTrustManagers();
+
+        try
+        {
+            SSLContext sslcontext = SSLContext.getInstance(TLS_PROTOCOL);
+            sslcontext.init(keyManagers, trustManagers, null);
+            return sslcontext;
+        }
+        catch(Throwable e)
+        {
+            throw new AlfrescoRuntimeException("Unable to create SSL context", e);
+        }
+    }
+
+    public static CloseableHttpClient createHttpClient(HttpClientConfig config)
+    {
+        HttpClientBuilder clientBuilder = HttpClients.custom();
+
+        if(Boolean.parseBoolean(config.getConfig().get("mTLSEnabled")))
+        {
+            clientBuilder.addInterceptorFirst((HttpRequestInterceptor) (request, context) -> {
+                if (!((HttpHost) context.getAttribute(HTTP_TARGET_HOST)).getSchemeName().equals(HTTPS_PROTOCOL))
+                {
+                    String msg = "mTLS is enabled but provided URL has not secured protocol";
+                    throw new HttpClientException(msg);
+                }
+            });
+
+            SSLConnectionSocketFactory sslConnectionSocketFactory = new SSLConnectionSocketFactory(
+                    createSSLContext(config),
+                    new String[] { TLS_V_1_2, TLS_V_1_3 },
+                    null,
+                    SSLConnectionSocketFactory.getDefaultHostnameVerifier());
+            clientBuilder.setSSLSocketFactory(sslConnectionSocketFactory);
+        }
+
+        clientBuilder.setMaxConnTotal(Integer.parseInt(config.getConfig().get("maxTotalConnections")));
```

Comment:
Consider applying [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter) here by hiding the internals of the config class. What about something like:
```java
config.getMaxTotalConnections()
```
instead of
```java
Integer.parseInt(config.getConfig().get("maxTotalConnections"))
```

Reply:
I had similar approach earlier, but I agreed with @mstrankowski to pass raw map end retreive properties directly in place of usage - such approach was considered as more beneficial if in the future someone will add more properties to config. But I can change that to yours suggestion @pzhyland

Reply:
So at least it could be change to something like this:
```java
config.getIntegerProperty("maxTotalConnections")
config.getBooleanProperty(...)
config.getKeyManagers()
config.getTrustManagers()
```

Reply:
That is a good idea. +1